### PR TITLE
Upgrade to PyTorch 2.0

### PR DESCRIPTION
**Pull request by mzweilin**
# What does this PR do?

Though the initial goal is to only upgrade PyTorch and `torchvision`, it's inevitable to upgrade `lightning` and `torchmetrics`, because the old `lightning` is not aware of the different API of LR scheduler in PyTorch 2.0, and the new `lightning` requires newer `torchmetrics`.

- Upgrade dependency.
- Adapt our code to the new dependency.

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] `pytest`
- [x] `CUDA_VISIBLE_DEVICES=0 python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu trainer.precision=16` reports 70% (21 sec/epoch).
- [x] `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 model.optimizer.lr=0.2 trainer.max_steps=2925 datamodule.ims_per_batch=256 datamodule.world_size=2` reports 70% (14 sec/epoch).

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/models/modular.py [hunk 1]
```diff
@@ -232,11 +232,15 @@ def log_metrics(self, metrics, prefix="", prog_bar=False):
         def enumerate_metric(metric, name):
             # Metrics can have arbitrary depth.
             if isinstance(metric, torch.Tensor):
-                metrics_dict[name] = metric
+                # Ignore non-scalar results generated by Metrics, such as list of classes from MAP.
```
Comment (dxoigmn) [inline mart/models/modular.py:235]: It would be nice to flatten so we get per class AP numbers. This doesn't work right now though and just returns the average 🙃

Comment (mzweilin) [reply mart/models/modular.py:235]: Good point. I would make per-class AP in a separate PR.

File: mart/attack/adversary.py [hunk 2]
```diff
@@ -172,8 +170,31 @@ def forward(self, *, model=None, sequence=None, **batch):
 
         return input_adv
 
-    def _attack(self, *, input, **batch):
+    # Make sure we can do autograd.
```
Comment (mzweilin) [inline mart/attack/adversary.py:173]: An alternative is `+trainer.inference_mode=false` for experiments that run adversarial robustness evaluation in val/test.

Comment (mzweilin) [reply mart/attack/adversary.py:173]: I don't like to turn off inference_mode for all experiments due to the performance advantage.

Comment (mzweilin) [reply mart/attack/adversary.py:177]: I would make `trainer.inference_mode=false` in the default setup, because this is an adversary library. I also keep the `de_inference()` method because it can be used by applications that runs in the inference mode.

Comment (mzweilin) [reply mart/attack/adversary.py]: I removed the `de_inference()` stuff that should live in a separate PR.

File: mart/models/modular.py [hunk 3]
```diff
@@ -145,9 +145,13 @@ def training_step(self, batch, batch_idx):
                 )
             self.training_metrics(output[self.output_preds_key], output[self.output_target_key])
 
-        return output[self.output_loss_key]
+        loss = output[self.output_loss_key]
```
Comment (dxoigmn) [inline mart/models/modular.py:148]: What happens if loss is included in training_step_log?

File: mart/models/modular.py [hunk 4]
```diff
@@ -232,11 +236,15 @@ def log_metrics(self, metrics, prefix="", prog_bar=False):
...
             else:
                 for k, v in metric.items():
                     enumerate_metric(v, f"{name}/{k}")
 
         enumerate_metric(metrics, prefix)
 
-        self.log_dict(metrics_dict, prog_bar=prog_bar)
+        # PyTorch Lightning: It is recommended to use `self.log('validation_metrics/acc', ..., sync_dist=True)`
+        # when logging on epoch level in distributed setting to accumulate the metric across devices.
+        self.log_dict(metrics_dict, prog_bar=prog_bar, sync_dist=True)
```
Comment (dxoigmn) [inline mart/models/modular.py:250]: I thought this wasn't necessary with torchmetrics?

File: mart/datamodules/modular.py [hunk 5]
```diff
@@ -29,6 +29,8 @@ def __init__(
         ims_per_batch=1,
         world_size=1,
         pin_memory=False,
+        # Classification metrics may require the value in config.
+        num_classes=None,
```
Comment (dxoigmn) [inline mart/datamodules/modular.py:33]: Why is this necessary if it is never used?

File: mart/models/modular.py (hunk 4 above)
Comment (mzweilin) [reply mart/models/modular.py]: You are right! https://torchmetrics.readthedocs.io/en/stable/pages/lightning.html

Fixed.

File: mart/datamodules/modular.py (hunk 5 above)
Comment (mzweilin) [reply mart/datamodules/modular.py:33]: The classification metrics in `torchmetrics` requires `num_classes`. I think it's best to specify it in the config of `datamodule` and reference to it in metrics. Otherwise, we have to specify it for every experiments.

File: mart/models/modular.py (hunk 3 above)
Comment (mzweilin) [reply mart/models/modular.py:148]: Then loss will be additionally logged as `training/loss` quietly (not on progress bar).

**Review by dxoigmn — APPROVED** [review]
LGTM!
